### PR TITLE
Circle Emitter

### PR DESCRIPTION
**Pull request by KZhu-ME**
Unit test is located in `mewarpx/tests/test_circle_emitter.py`

### CODE REVIEW

File: Python/pywarpx/_libwarpx.py [hunk 1]
```diff
@@ -1110,6 +1110,8 @@ def get_mesh_current_density_fp_pml(level, direction, include_ghosts=True):
         return _get_mesh_field_list(libwarpx.warpx_getCurrentDensityFP_PML, level, direction, include_ghosts)
     except ValueError:
         raise Exception('PML not initialized')
+
+
```
Comment (roelof-groenewald) [inline Python/pywarpx/_libwarpx.py]: For all the changes to this file: I understand the desire for neatness but since this file is present in the upstream repo these changes only serve to diverge our fork from upstream without adding any functionality. You can submit an upstream PR for these changes if you like but we shouldn't have this be changed only in our repo.

File: mewarpx/examples/capacitive_discharge.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 
 from pywarpx import picmi
 import pywarpx
+from pywarpx import callbacks
```
Comment (roelof-groenewald) [inline mewarpx/examples/capacitive_discharge.py]: Why was this added?

File: mewarpx/mewarpx/assemblies.py [hunk 3]
```diff
@@ -93,22 +89,63 @@ def __init__(self, center_x, center_z, radius, V, T, WF, name,
             T (float): Temperature (K)
             WF (float): Work function (eV)
             name (str): Assembly name
-            install_in_fieldsolver (float): If True and the Assembly is an
+            install_in_fieldsolver (bool): If True and the Assembly is an
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/assemblies.py]: Please remember to change this to `install_in_simulation`.

File: mewarpx/mewarpx/assemblies.py [hunk 4]
```diff
@@ -93,22 +89,63 @@ def __init__(self, center_x, center_z, radius, V, T, WF, name,
...
                 embedded boundary it will be included in the WarpX fieldsolver
         """
         super(Cylinder, self).__init__(V=V, T=T, WF=WF, name=name)
-
+        # Y is always treated as the height
         self.center_x = center_x
         self.center_z = center_z
+        self.height = 1 # m
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/assemblies.py]: As far as I can tell this is not used anywhere, specifically not in the `isinside` function or the implicit function. So the physical object actually has y = infinity. Can we therefore remove it?

File: mewarpx/mewarpx/emission.py [hunk 5]
```diff
@@ -0,0 +1,1072 @@
...
+    """ ArbitraryEmitter2D class takes in a conductor, calculates an approximate
+    surface that encloses the conductor and then sets up the appropriate
+    emitting surfaces, given a number of particles to emit.
+    """
+    geoms = ['XZ']
+
+    def __init__(self, conductor, T, res_fac=5., transverse_fac=1.0, **kwargs):
+        """Construct the emitter based on conductor object and temperature.
+        Arguments:
+            conductor (warp.Assembly object): Conductor to emit from.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py]: Update docstring - no longer a `warp.Assembly` object.

**Review by roelof-groenewald — CHANGES_REQUESTED** [review]
I left a couple comments for mostly minor things.

File: mewarpx/mewarpx/util.py [hunk 6]
```diff
@@ -5,9 +5,26 @@
 import inspect
 import os
 
+import numpy as np
+import minerva.util as minutil
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/util.py]: I believe this will already be fixed in @nruof's PR but it would be good to confirm that we no longer need `minerva` once you merge in his changes (might not be on github yet).

File: mewarpx/examples/capacitive_discharge.py (hunk 2 above)
Comment (nruof) [reply mewarpx/examples/capacitive_discharge.py]: It'll go away when nruof/thermionic_emission branch is merged

File: mewarpx/mewarpx/util.py (hunk 6 above)
Comment (nruof) [reply mewarpx/mewarpx/util.py]: Yes

File: mewarpx/mewarpx/assemblies.py [hunk 7]
```diff
@@ -93,22 +89,63 @@ def __init__(self, center_x, center_z, radius, V, T, WF, name,
...
+                units of meters.
+
+        Returns:
+            result (np.ndarray): array of flattened grid where all tiles
+                inside the cylinder are 1, and other tiles are 0.
+        """
+        result = np.where((X - self.center_x)**2 + (Z - self.center_z)**2 <= (self.radius + aura)**2,
+                             1, 0)
+
+        return result
```
Comment (nruof) [inline mewarpx/mewarpx/assemblies.py]:
```suggestion
dist2 = (X - self.center_x)**2 + (Z - self.center_z)**2
boundary2 = (self.radius + aura)**2

return np.where(dist2 <= boundary2, 1, 0)
```

File: Python/pywarpx/_libwarpx.py (hunk 1 above)
Comment (KZhu-ME) [reply Python/pywarpx/_libwarpx.py]: That makes sense, I'll revert this change. But the changes to get_gathered_phi_fp and others can stay right, as these functions are not upstream anyway?

File: mewarpx/mewarpx/util.py (hunk 6 above)
Comment (KZhu-ME) [reply mewarpx/mewarpx/util.py]: Sounds good, I'll change the references to use constants instead of minutil

**Comment by KZhu-ME** [comment]
Here are files with the data of particles generated at the circle emitter from warp and warpx. The x,y,z positions seem to match very closely, but the velocities, vx, vy, and vz of the warpx results seem to skew lower compared to the warp results (in the range of +/- 750000 against warp's +/- 850000). However, the standard deviations also match very closely.

[warp_emission.csv](https://github.com/ModernElectron/WarpX/files/6851652/warp_emission.csv)
[warpx_emission.csv](https://github.com/ModernElectron/WarpX/files/6851653/warpx_emission.csv)

File: Source/Evolve/WarpXEvolve.cpp [hunk 8]
```diff
@@ -165,6 +165,7 @@ WarpX::Evolve (int numsteps)
         if (do_electrostatic != ElectrostaticSolverAlgo::None)
         {
             const bool skip_deposition = true;
+            if (warpx_py_particleinjection) warpx_py_particleinjection();
```
Comment (roelof-groenewald) [inline Source/Evolve/WarpXEvolve.cpp]: I believe this should now be removed since I think it is duplicated a few lines earlier.

File: mewarpx/mewarpx/util.py [hunk 9]
```diff
@@ -156,16 +158,119 @@ def mkdir_p(path):
...
+    # Random x and y positions
+    x = xmin + (xmax - xmin) * np.random.rand(num_samples)
+    y = ymin + (ymax - ymin) * np.random.rand(num_samples)
+    z = np.ones_like(x) * z
+
+    if rseed is not None:
+        np.random.set_state(nprstate)
+
+    return x, y, z
     return (p * constants.torr_cgs) / (constants.kb_cgs * T)
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/util.py]: I'm guessing this is due to some merging issue?
```suggestion
```

File: mewarpx/tests/test_circle_emitter.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    )
+
+    # MCC Collisions
+    MCC(
+        electron_species=electrons,
+        ion_species=ions,
+        T_INERT=T_INERT,
+        N_INERT=N_INERT,
+        exclude_collisions=['charge_exchange'])
+
```
Comment (roelof-groenewald) [inline mewarpx/tests/test_circle_emitter.py]:
```suggestion
        exclude_collisions=['charge_exchange']
    )

```

File: mewarpx/tests/test_circle_emitter.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    D_CA = 1 # m
+    V_bias = 0 # V
+
+    ###################################
+    # numerics parameters
+    ##################################
+    # Initialize each run with consistent, randomly-chosen, rseed. Use a random
+    # seed instead for initial dataframe generation.
+    np.random.seed(671237741)
+
```
Comment (roelof-groenewald) [inline mewarpx/tests/test_circle_emitter.py:32]:
```suggestion
```

File: mewarpx/mewarpx/util.py [hunk 12]
```diff
@@ -156,16 +158,119 @@ def mkdir_p(path):
...
+        np.random.set_state(nprstate)
+    return vx, vy, vz
+
+
+def get_positions(num_samples, xmin, xmax, ymin=0, ymax=0, z=0,
+                rseed=None):
+
+    """Provide random samples of [x, y, z] for electrons in simulation.
+    In x and y, positions are uniformly distributed. In z, positions are
+    placed at the emitter.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/util.py:249]:
```suggestion
    placed at the emitter.
    
```

File: mewarpx/mewarpx/util.py [hunk 13]
```diff
@@ -156,16 +158,119 @@ def mkdir_p(path):
...
+    Arguments:
+        num_samples (int): Number of particles to generate positions for
+        xmin (float): Min position in x (meters)
+        xmax (float): Max position in x (meters)
+        ymin (float): Min position in y (meters)
+        ymax (float): Max position in y (meters)
+        z (float): Position of the emitter on the z-axis (meters)
+        rseed (positive int): If specified, seed the random number generator.
+            Used for testing. The random number generator is set back at the
+            end of the function.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/util.py:260]:
```suggestion
            end of the function.
            
```

File: mewarpx/mewarpx/util.py [hunk 14]
```diff
@@ -156,16 +158,119 @@ def mkdir_p(path):
...
-        N (float): Number density of gas atoms/molecules (1/cm^3)
+        N (float): Number density of gas atoms/molecules (1/m^3)
+    """
+    return (p * constants.torr_cgs) / (constants.kb_cgs * T) * 1e6
+
+
+def get_velocities(num_samples, T, m, emission_type='thermionic',
+                    transverse_fac=1.0, rseed=None):
+
+    """Generate array of random [vx, vy, vz] for cathode-emitted electrons.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/util.py:177]:
```suggestion
    """Generate array of random [vx, vy, vz] for cathode-emitted electrons.
    
```

File: mewarpx/mewarpx/util.py [hunk 15]
```diff
@@ -156,16 +158,119 @@ def mkdir_p(path):
...
+            along +zhat, and use "random" for a purely thermal distribution
+            with no preferred direction. "half_maxwellian" is used at present
+            for surface ionization, again along +zhat. Defaults to
+            "thermionic".
+        transverse_fac (float): Scale the particle's x and y average energies
+            by this factor, scales z average energy to conserve total average
+            particle energy in the distribution. Default 1., Min 0., Max 2.
+        rseed (positive int): If specified, seed the random number generator.
+            Used for testing. The random number generator is set back at the
+            end of the function.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/util.py:193]:
```suggestion
            end of the function.
            
```

File: mewarpx/mewarpx/emission.py [hunk 16]
```diff
@@ -844,7 +851,188 @@ def _get_xv_coords(self, npart, m, rseed):
...
+class ArbitraryEmitter2D(Emitter):
+
+    """ ArbitraryEmitter2D class takes in a conductor, calculates an approximate
+    surface that encloses the conductor and then sets up the appropriate
+    emitting surfaces, given a number of particles to emit.
+    """
+    geoms = ['XZ']
 
+    def __init__(self, conductor, T, res_fac=5., transverse_fac=1.0, **kwargs):
+        """Construct the emitter based on conductor object and temperature.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py:878]:
```suggestion
        """Construct the emitter based on conductor object and temperature.
        
```

File: mewarpx/mewarpx/emission.py [hunk 17]
```diff
@@ -844,7 +851,188 @@ def _get_xv_coords(self, npart, m, rseed):
 
     def get_normals(self, x, y, z):
         """Calculate local surface normal at specified coordinates.
+        Arguments:
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py]:
```suggestion
        
        Arguments:
```

File: mewarpx/mewarpx/assemblies.py [hunk 18]
```diff
@@ -93,23 +93,62 @@ def __init__(self, center_x, center_z, radius, V, T, WF, name,
...
+        dist_to_center = (X - self.center_x)**2 + (Z - self.center_z)**2
+        boundary = (self.radius + aura)**2
+        result = np.where(dist_to_center <= boundary, 1, 0)
+
+        return result
+
+    def calculatenormal(self, px, py, pz):
+        """
+        Calculates Normal of particle inside/outside of conductor to nearest
+        surface of conductor
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/assemblies.py:135]:
```suggestion
        surface of conductor
        
```

File: mewarpx/mewarpx/assemblies.py [hunk 19]
```diff
@@ -93,23 +93,62 @@ def __init__(self, center_x, center_z, radius, V, T, WF, name,
...
+        Arguments:
+            px (np.ndarray): x-coordinate of particle in meters.
+            py (np.ndarray): y-coordinate of particle in meters.
+            pz (np.ndarray): z-coordinate of particle in meters.
+        Returns:
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/assemblies.py:142]:
```suggestion
        
        Returns:
```

File: mewarpx/mewarpx/assemblies.py [hunk 20]
```diff
@@ -93,23 +93,62 @@ def __init__(self, center_x, center_z, radius, V, T, WF, name,
             T (float): Temperature (K)
             WF (float): Work function (eV)
             name (str): Assembly name
-            install_in_fieldsolver (float): If True and the Assembly is an
+            install_in_simulation (bool): If True and the Assembly is an
                 embedded boundary it will be included in the WarpX fieldsolver
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/assemblies.py]:
```suggestion
                embedded boundary it will be included in the WarpX simulation
```

**Review by roelof-groenewald — CHANGES_REQUESTED** [review]
Looks good. I left a couple style change suggestions. Besides that the circleCI tests are failing since Numba isn't installed - please add that as a requirement.

File: mewarpx/mewarpx/emission.py [hunk 21]
```diff
@@ -844,7 +851,188 @@ def _get_xv_coords(self, npart, m, rseed):
...
+            x (np.ndarray): x-coordinates of emitted particles (in meters).
+            y (np.ndarray): y-coordinates of emitted particles (in meters).
+            z (np.ndarray): z-coordinates of emitted particles (in meters).
+        Returns:
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/emission.py]:
```suggestion

        Returns:
```

File: mewarpx/mewarpx/util.py (hunk 9 above)
Comment (KZhu-ME) [reply mewarpx/mewarpx/util.py]: Oops, yeah good catch, this was from merging.

**Review by roelof-groenewald — APPROVED** [review]
Nice work!

**Review by nruof — APPROVED** [review]
